schemes/musica/musica_ccpp_species.F90

```diff
@@ -0,0 +1,322 @@
+      end if
+      indices_constituent_props(i_elem) = index_species
+    end do
+
+  end subroutine find_musica_species_indices
+
+  !> Initializes arrays to store the species indices of the CCPP constituents
+  subroutine initialize_musica_species_indices(constituent_props, errmsg, errcode)
+    use ccpp_constituent_prop_mod, only: ccpp_constituent_prop_ptr_t
+
+    type(ccpp_constituent_prop_ptr_t), intent(in)  :: constituent_props(:)
+    character(len=512),                intent(out) :: errmsg
+    integer,                           intent(out) :: errcode
+
+    if (.not. allocated( micm_species_set ) .or. &
+        .not. allocated( tuvx_species_set )) then
+      errmsg = "[MUSICA Error] The MUSICA species set(s) are not allocated."
+      errcode = 1
+      return
+    end if
+
+    allocate( micm_indices_constituent_props( size(micm_species_set) ) )
+    call find_musica_species_indices( constituent_props, micm_species_set, &
+                          micm_indices_constituent_props, errmsg, errcode )
+    if (errcode /= 0) return
+
+    allocate( tuvx_indices_constituent_props( size(tuvx_species_set) ) )
+    call find_musica_species_indices( constituent_props, tuvx_species_set, &
+                          tuvx_indices_constituent_props, errmsg, errcode )
+    if (errcode /= 0) return
+
+  end subroutine initialize_musica_species_indices
+
+  !> Iterates through the constituent property pointer array to populate the molar mass array,
+  ! storing molar mass values in a sequence that matches the order of the MICM state array
+  subroutine initialize_molar_mass_array(constituent_props, errmsg, errcode)
+    use ccpp_constituent_prop_mod, only: ccpp_constituent_prop_ptr_t
+
+    type(ccpp_constituent_prop_ptr_t), intent(in)  :: constituent_props(:)
+    character(len=512),                intent(out) :: errmsg
+    integer,                           intent(out) :: errcode
+
+    ! local variables
+    integer :: i_elem, index_species
+
+    if (.not. allocated( micm_species_set )) then
+      errmsg = "[MUSICA Error] The MICM species set is not allocated."
+      errcode = 1
+      return
+    end if
+
+    allocate( micm_molar_mass_array( size(micm_species_set) ) )
+    do i_elem = 1, size(micm_species_set)
+      call constituent_props( micm_species_set(i_elem)%index_constituent_props ) &
+                            %molar_mass( micm_molar_mass_array(i_elem), errcode, errmsg )
+      if (errcode /= 0) then
+        errmsg = "[MUSICA Error] Unable to get molar mass for " // micm_species_set(i_elem)%name
+        return
+      end if
+    end do
+
+    ! TODO(jiwon) Check molar mass is non zero as it becomes a denominator for unit converison
+    do i_elem = 1, size(micm_molar_mass_array)
+      if (micm_molar_mass_array(i_elem) <= 0) then
+        errcode = 1
+        errmsg = "[MUSICA Error] Molar mass must be greater than zero for " &
+                  // micm_species_set(i_elem)%name
+        return
+      end if
+    end do
```

Comment:
Hi @peverwhee, the last time we discussed this, we mentioned that the check could be done in the CCPP framework. Just wondering, do we have any updates on this?

Reply:
hi @boulderdaze. Nope! I just made this issue - https://github.com/NCAR/ccpp-framework/issues/620 - to keep track of this and the other plan to check if it has been set.

Not sure when I'll be able to get to it, but at least we're tracking it!

Reply:
Sounds good, thank you!